libnavigation-core/src/androidTest/java/com/mapbox/navigation/core/trip/service/ArtificialDriverTest.kt

```diff
@@ -0,0 +1,143 @@
+package com.mapbox.navigation.core.trip.service
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.platform.app.InstrumentationRegistry
+import com.mapbox.navigation.base.options.NavigationOptions
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.route.RouterOrigin
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.MapboxNavigationProvider
+import com.mapbox.navigation.core.navigator.toFixLocation
+import com.mapbox.navigation.core.replay.history.ReplayEventUpdateLocation
+import com.mapbox.navigation.core.replay.history.mapToLocation
+import com.mapbox.navigation.core.replay.route.ReplayRouteMapper
+import com.mapbox.navigation.core.test.R
+import com.mapbox.navigation.navigator.internal.MapboxNativeNavigator
+import com.mapbox.navigation.navigator.internal.MapboxNativeNavigatorImpl
+import com.mapbox.navigator.NavigationStatus
+import com.mapbox.navigator.NavigationStatusOrigin
+import com.mapbox.navigator.NavigatorObserver
+import com.mapbox.navigator.RouteState
+import com.mapbox.navigator.SetRoutesReason
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.async
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.takeWhile
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.runBlocking
+import org.junit.Assert.assertTrue
+import org.junit.Ignore
+import org.junit.Test
+import org.junit.runner.RunWith
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+@RunWith(AndroidJUnit4::class)
+class ArtificialDriverTest {
+
+    @Test
+    @Ignore("test sometimes fails because of https://mapbox.atlassian.net/browse/NN-418")
+    fun nativeNavigatorFollowsArtificialDriverWithoutReroutes() =
+        runBlocking<Unit>(Dispatchers.Main) {
+            withNavigators { mapboxNavigation, nativeNavigator ->
+                mapboxNavigation.historyRecorder.startRecording()
+                val testRoute = getTestRoute()
+                val events = createArtificialLocationUpdates(testRoute)
+                val setRoutesResult =
+                    nativeNavigator.setRoutes(testRoute, reason = SetRoutesReason.NEW_ROUTE)
+                assertTrue("result is $setRoutesResult", setRoutesResult.isValue)
+                val statusesTracking = async<List<NavigationStatus>> {
+                    nativeNavigator.collectStatuses(untilRouteState = RouteState.COMPLETE)
+                }
+
+                for (location in events.map { it.location.mapToLocation() }) {
+                    assertTrue(nativeNavigator.updateLocation(location.toFixLocation()))
+                }
```

Comment:
Like history recording replay, you can take the locations with proper monotonic time in them and feed back-to-back into the navigator - it will produce the correct result.